api.md

```diff
@@ -0,0 +1,215 @@
+
+General JSON guidelines:
+
+* Responses should be **a JSON object** (not an array). Using an array to return results limits the ability to include metadata about results, and limits the API's ability to add additional top-level keys in the future.
+* **Don't use unpredictable keys**. Parsing a JSON response where keys are unpredictable (e.g. derived from data) is difficult, and adds friction for clients.
+* **Use consistent case for keys**. Whether you use `under_score` or `CamelCase` for your API keys, make sure you are consistent.
+
+### Use a consistent date format
+
+And specifically, [use ISO 8601](https://xkcd.com/1179/), in UTC.
+
+For just dates, that looks like `2013-02-27`. For full times, that's of the form `2013-02-27T10:00:00Z`.
+
+This date format is used all over the web, and puts each field in consistent order -- from least granular to most granular.
+
+
+### API Keys
+
+These standards do not take a position on whether or not to use API keys.
+
+But _if_ keys are used to manage and authenticate API access, the API should allow some sort of unauthenticated access, without keys.
+
+This allows newcomers to use and experiment with the API in demo environments and with simple `curl`/`wget`/etc. requests.
+
+Consider whether one of your product goals is to allow a certain level of normal production use of the API without enforcing advanced registration by clients.
+
+
+### Error handling
+
+Handle all errors (including otherwise uncaught exceptions) and return a data structure in the same format as the rest of the API.
+
+For example, a JSON API might provide the following when an uncaught exception occurs:
+
+```json
+{
+  "message": "Description of the error.",
+  "exception": "[detailed stacktrace]"
+}
+```
+
+HTTP responses with error details should use a `4XX` status code to indicate a client-side failure (such as invalid authorization, or an invalid parameter), and a `5XX` status code to indicate server-side failure (such as an uncaught exception).
+
+
+### Pagination
+
+If pagination is required to navigate datasets, use the method that makes the most sense for the API's data.
+
+#### Parameters
+
+Common patterns:
+
+* `page` and `per_page`. Intuitive for many use cases. Links to "page 2" may not always contain the same data.
+* `offset` and `limit`. This standard comes from the SQL database world, and is a good option when you need stable permalinks to result sets.
+* `since` and `limit`. Get everything "since" some ID or timestamp. Useful when it's a priority to let clients efficiently stay "in sync" with data. Generally requires result set order to be very stable.
+
+#### Metadata
+
+Include enough metadata so that clients can calculate how much data there is, and how and whether to fetch the next set of results.
+
+Example of how that might be implemented:
+
+```json
+{
+  "results": [ ... actual results ... ],
+  "pagination": {
+    "count": 2340,
+    "page": 4,
+    "per_page": 20
+  }
+}
+```
```

Comment:
I really, really don't like this pattern. I'd much rather have a link to following pages than have to calculate the size of the dataset and then construct a URL for the next page myself.